core/base/batch_struct.hpp

```diff
@@ -0,0 +1,144 @@
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_CORE_BASE_BATCH_STRUCT_HPP_
+#define GKO_CORE_BASE_BATCH_STRUCT_HPP_
+
+
+#include <ginkgo/core/base/array.hpp>
+#include <ginkgo/core/base/lin_op.hpp>
+#include <ginkgo/core/base/types.hpp>
+
+
+namespace gko {
+namespace batch_multi_vector {
+
+
+/**
+ * Encapsulates one matrix from a batch of dense matrices (vectors).
+ */
+template <typename ValueType>
+struct BatchEntry {
+    using value_type = ValueType;
+    ValueType* values;
+    size_type stride;
+    int num_rows;
+    int num_rhs;
+};
+
+/**
+ * A 'simple' structure to store a global uniform batch of dense matrices.
+ *
+ * It is uniform in the sense that all matrices in the batch have common sizes.
+ */
+template <typename ValueType>
+struct UniformBatch {
+    using value_type = ValueType;
+    using entry_type = BatchEntry<ValueType>;
+
+    ValueType* values;
+    size_type num_batch_entries;
+    size_type stride;
+    int num_rows;
+    int num_rhs;
+
+    size_type get_entry_storage() const
+    {
+        return num_rows * stride * sizeof(value_type);
+    }
+};
+
+
+}  // namespace batch_multi_vector
+
+
+namespace batch {
+
+
+template <typename ValueType>
+GKO_ATTRIBUTES GKO_INLINE gko::batch_multi_vector::BatchEntry<const ValueType>
+to_const(const gko::batch_multi_vector::BatchEntry<ValueType>& b)
+{
+    return {b.values, b.stride, b.num_rows, b.num_rhs};
+}
+
+
+template <typename ValueType>
+GKO_ATTRIBUTES GKO_INLINE gko::batch_multi_vector::UniformBatch<const ValueType>
+to_const(const gko::batch_multi_vector::UniformBatch<ValueType>& ub)
+{
+    return {ub.values, ub.num_batch_entries, ub.stride, ub.num_rows,
+            ub.num_rhs};
+}
+
+
+/**
+ * Extract one object (matrix, vector etc.) from a batch of objects
+ *
+ * This overload is for batch multi-vectors.
+ * These overloads are intended to be called from within a kernel.
+ *
+ * @param batch  The batch of objects to extract from
+ * @param batch_idx  The position of the desired object in the batch
+ */
+template <typename ValueType>
+GKO_ATTRIBUTES GKO_INLINE batch_multi_vector::BatchEntry<ValueType> batch_entry(
```

Comment:
`entry` seems ambiguous to me, how about `tile`?

Reply:
I feel `tile` is more confusing as we are handling one batch entry here. You can again have tiles inside each batch entry in the context of parallelism.